data/posts/react-router-and-react-server-components.md

```diff
@@ -0,0 +1,149 @@
+
+## How We Handle Server Challenges Today
+
+Let's look at how React Router Framework Mode and React Server Components each tackle these common problems:
+
+### Inlining Data
+
+**Framework Mode**: You return data from loader functions, and React Router handles getting it to your components.
+
+```js
+// Framework Mode
+export async function loader() {
+  const user = await getUser();
+  return { user };
+}
+
+export default function Profile() {
+  const { user } = useLoaderData();
+  return <h1>{user.name}</h1>;
+}
+```
+
+**React RSC**: You pass props to `"use client"` components from server components.
+
+```js
+// React RSC
+async function ProfileServer() {
+  const user = await getUser();
+  return <ProfileClient user={user} />;
+}
+```
+
+**React Router RSC**: You get both options! Use whichever pattern fits your needs.
+
+### Streaming UI
+
+**Framework Mode**: Return promises from loaders and use the `<Await>` component.
+
+```js
+// Framework Mode
+export function loader() {
+  return defer({
+    criticalData: getCriticalData(),
+    slowData: getSlowData(), // This is a promise
+  });
+}
+
+export default function Page() {
+  const { criticalData, slowData } = useLoaderData();
+  return (
+    <>
+      <h1>{criticalData.title}</h1>
+      <Suspense fallback={<Spinner />}>
+        <Await resolve={slowData}>
+          {(data) => <SlowContent data={data} />}
+        </Await>
+      </Suspense>
+    </>
+  );
+}
+```
+
+**React RSC**: Pass promises as props and use `await` on the server and `use(promise)` on the client.
+
+**React Router RSC**: Again, both patterns are available to you.
+
+### Code Splitting Routes
+
+**Framework Mode**: A `routes.ts` config file processed by a bundler plugin generates a manifest that maps routes to chunks.
+
+**React RSC**: Dynamic imports on the server combined with `"use client"` directives mean the browser only downloads what it needs.
+
+## Under the Hood: How Framework Mode Works
+
+React Router's Framework Mode is powered by some clever bundler integration:
+
+1. You define routes in a `routes.ts` config file
+2. A bundler plugin reads this config and generates:
+   - A manifest mapping routes to chunks for the browser
+   - A server entry containing the manifest and server-side route mappings
+3. The React Router runtime uses these generated files to handle routing
+
+This approach has served us well, but it requires significant bundler integration and complexity.
+
+## The New Way: React Router RSC
+
+React Router RSC takes a different approach that's actually simpler:
+
+- Routes defined in app code like any other data structure (no more special `routes.ts` bundler plugins or build-time generation required)
+- A function that handles routing HTML vs data requests
+- A function that generates RSC payloads that get rendered to HTML somewhere
```

Comment:
Trying to match the tense of the bullets above so it's easier to compare and contrast

I will say, the second 2 lines feel a bit hand-wavy. It could just be my unfamiliarity, but I worry that these sentences don't necessarily make it sound "simpler". Are you saying it's simpler because a function is handling this rather than a bundler/file generation?

```suggestion
- You define routes in app code like any other data structure (no more special `routes.ts` bundler plugins or build-time generation required)
- A function handles routing HTML vs data requests
- A function generates RSC payloads that get rendered to HTML somewhere
```